pkg/agent/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,451 @@
+		return true
+	} else if err := c.syncService(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Service %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *ServiceExternalIPController) deleteService(service apimachinerytypes.NamespacedName) error {
+	c.externalIPStatesMutex.Lock()
+	defer c.externalIPStatesMutex.Unlock()
+	if state, exist := c.externalIPStates[service]; !exist {
+		return nil
+	} else {
+		delete(c.externalIPStates, service)
+		return c.ipAssigner.UnassignIP(state.ip)
+	}
+}
+
+func (c *ServiceExternalIPController) getServiceState(service *corev1.Service) (externalIPState, bool) {
+	c.externalIPStatesMutex.RLock()
+	defer c.externalIPStatesMutex.RUnlock()
+	name := apimachinerytypes.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}
+	state, exist := c.externalIPStates[name]
+	return state, exist
+}
+
+func (c *ServiceExternalIPController) saveServiceState(service *corev1.Service, state externalIPState) {
+	c.externalIPStatesMutex.Lock()
+	defer c.externalIPStatesMutex.Unlock()
+	name := apimachinerytypes.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}
+	c.externalIPStates[name] = state
+}
+
+func (c *ServiceExternalIPController) getServiceExternalIP(service *corev1.Service) string {
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return ""
+	}
+	return service.Status.LoadBalancer.Ingress[0].IP
+}
+
+func (c *ServiceExternalIPController) syncService(key apimachinerytypes.NamespacedName) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Service for %s. (%v)", key, time.Since(startTime))
+	}()
+
+	service, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return c.deleteService(key)
+		}
+		return err
+	}
+
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return c.deleteService(key)
+	}
+
+	state, exist := c.getServiceState(service)
+	currentExternalIP := c.getServiceExternalIP(service)
+	if exist && state.ip != currentExternalIP {
+		if err := c.deleteService(key); err != nil {
+			return err
+		}
+	}
+
+	ipPoool := service.ObjectMeta.Annotations[types.ExternalIPPoolAnnotationKey]
+	if currentExternalIP == "" || ipPoool == "" {
+		return nil
+	}
+
+	selectNode := true
+	var filters []func(string) bool
+	if service.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+		nodes, err := c.nodesHasHealthyServiceEndpoint(service)
+		if err != nil {
+			return err
+		}
+		// Avoid unnecessary migration caused by Endpoint changes.
+		if exist && c.cluster.AliveNodes().Has(state.assignedNode) && nodes.Has(state.assignedNode) {
```

Comment:
Will this really get consistent result from multiple agents?
For example, restart any agent, it will re-caculate the node of a service based on consistentHash while other agents may keep using a different Node selected previously?

Reply:
It did will happen if an agent restarts and promote it as the owner. The cache might differ among Nodes if agent restarts but not selected as the owner of the IP. But I think It will be eventually constitant if the winer promote itself and update she status of the Service. I think it is still better than possibly selecting a new Node everytime when endpoint get updated. What do you think?

Reply:
I agree to keep the Node when it has a valid endpoint. However, I don't understand why it behaves differently when a agent restarts. I think it's better to stick one strategy across multiple Nodes. And I don't find how the state will be eventually consistent: from the code, if the restarted agent doesn't select itself as the owner, the Node kept in cache will always be different from other agents, I'm not sure what's the consequence but this is not risky.
Couldn't we follow the same rule for a restarted agent? i.e. if the current owner has a valid endpoint, don't try to select a new owner?